include/alpaka/mem/buf/Traits.hpp

```diff
@@ -191,4 +191,64 @@ namespace alpaka
 
         ALPAKA_UNREACHABLE(allocBuf<TElem, TIdx>(host, extent));
     }
+
+    namespace detail
+    {
+        template<typename DevDst, typename DevSrc>
+        auto canZeroCopy(const DevDst& devDst, const DevSrc& devSrc) -> bool
+        {
+            if constexpr(std::is_same_v<DevDst, DevSrc>)
+                if(devSrc == devDst)
+                    return true;
+            return false;
+        }
+    } // namespace detail
+
+    //! Makes the content of the source view available on the device associated with the destination queue. If the
+    //! destination shares the same memory space as the source view, no copy is performed and the destination view is
+    //! updated to share the same buffer as the source view. Otherwise, a memcpy is performed.
+    template<typename TQueue, typename TViewDst, typename TViewSrc>
+    ALPAKA_FN_HOST void zero_memcpy(TQueue& dstQueue, TViewDst& viewDstOpt, TViewSrc const& viewSrc)
+    {
+        ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
+
+        if(detail::canZeroCopy(getDev(dstQueue), getDev(viewSrc)))
+        {
+#if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+            std::cout << "zero_memcopy: copy elided\n";
+#endif
+            viewDstOpt = viewSrc;
+        }
+        else
+        {
+#if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+            std::cout << "zero_memcopy: deep copy required\n";
+#endif
+            memcpy(dstQueue, viewDstOpt, viewSrc);
+        }
+    }
+
+    //! Makes the content of the source view available on the device associated with the destination queue. If the
+    //! destination shares the same memory space as the source view, no copy is performed and the source view is
+    //! returned. Otherwise
+    template<typename TQueue, typename TViewSrc>
+    ALPAKA_FN_HOST auto zero_memcpy(TQueue& dstQueue, TViewSrc const& viewSrc)
```

Comment:
This version is used when the user does not have a destination buffer yet. Typically, the case when transferring from the host to a (potentially newly allocated) device buffer.

Reply:
And I had to add another overload so the user could specify the destination device, because this is not necessarily the same as the queue device.